neps/nep-0568.md

```diff
@@ -0,0 +1,629 @@
+
+To enable efficient shard state management during resharding, Resharding V3 uses the `DBCol::ShardUIdMapping` column. This mapping allows child shards to reference ancestor shard data, avoiding the need for immediate duplication of state entries.
+
+#### Mapping Application in Adapters
+
+The core of the mapping logic is applied in `TrieStoreAdapter` and `TrieStoreUpdateAdapter`, which act as layers over the general `Store` interface. Here’s a breakdown of the key functions involved:
+
+* **Key Resolution**:
+
+  The `get_key_from_shard_uid_and_hash` function is central to determining the correct `ShardUId` for state access. At a high level, operations use the child shard's `ShardUId`, but within this function, the `DBCol::ShardUIdMapping` column is checked to determine if an ancestor `ShardUId` should be used instead.
+
+  ```rust
+  fn get_key_from_shard_uid_and_hash(
+      store: &Store,
+      shard_uid: ShardUId,
+      hash: &CryptoHash,
+  ) -> [u8; 40] {
+      let mapped_shard_uid = store
+          .get_ser::<ShardUId>(DBCol::StateShardUIdMapping, &shard_uid.to_bytes())
+          .expect("get_key_from_shard_uid_and_hash() failed")
+          .unwrap_or(shard_uid);
+      let mut key = [0; 40];
+      key[0..8].copy_from_slice(&mapped_shard_uid.to_bytes());
+      key[8..].copy_from_slice(hash.as_ref());
+      key
+  }
+  ```
+
+  This function first attempts to retrieve a mapped ancestor `ShardUId` from `DBCol::ShardUIdMapping`. If no mapping exists, it defaults to the provided child `ShardUId`. This resolved `ShardUId` is then combined with the `node_hash` to form the final key used in `State` column operations.
+
+* **State Access Operations**:
+
+  The `TrieStoreAdapter` and `TrieStoreUpdateAdapter` use `get_key_from_shard_uid_and_hash` to correctly resolve the key for both reads and writes. Example methods include:
+
+  ```rust
+  // In TrieStoreAdapter
+  pub fn get(&self, shard_uid: ShardUId, hash: &CryptoHash) -> Result<Arc<[u8]>, StorageError> {
+      let key = get_key_from_shard_uid_and_hash(self.store, shard_uid, hash);
+      self.store.get(DBCol::State, &key)
+  }
+
+  // In TrieStoreUpdateAdapter
+  pub fn increment_refcount_by(
+      &mut self,
+      shard_uid: ShardUId,
+      hash: &CryptoHash,
+      data: &[u8],
+      increment: NonZero<u32>,
+  ) {
+      let key = get_key_from_shard_uid_and_hash(self.store, shard_uid, hash);
+      self.store_update.increment_refcount_by(DBCol::State, key.as_ref(), data, increment);
+  }
+  ```
+
+  The `get` function retrieves data using the resolved `ShardUId` and key, while `increment_refcount_by` manages reference counts, ensuring correct tracking even when accessing data under an ancestor shard.
+
+#### Mapping Retention and Cleanup
+
+Mappings in `DBCol::ShardUIdMapping` persist as long as any descendant relies on an ancestor’s data. To manage this, the `set_shard_uid_mapping` function in `TrieStoreUpdateAdapter` adds a new mapping during resharding:
+
+```rust
```

Comment:
If I understand this mapping correctly, it is applied to all keys for the child shards, not just bridging to historical data. By which I mean even new keys for the child shard will be stored with the ancestor's shard UID.

I think this means the mapping needs to be taken into account when assigning the child. For example, suppose shard A splits into B and C, then shard B splits into D and E. In this case, D and E need to map to A instead of B because there are not any actual B keys in the database in the first place. Is that right or have I misunderstood?

Reply:
@Trisfald Can you reply please?

Reply:
> even new keys for the child shard will be stored with the ancestor's shard UID.

Yes.



> suppose shard A splits into B and C, then shard B splits into D and E. In this case, D and E need to map to A instead of B because there are not any actual B keys in the database in the first place. Is that right or have I misunderstood?

That's right if a node did two splits.
Another case would be a node did state sync after resharding and downloaded the state of shard B directly. Then it won't have the mapping from B to A.

Reply:
Thanks for the clarification @staffik ! Should the code block be updated so that the `parent_shard_uid` is replaced with it's mapped value (if present)?

> Another case would be a node did state sync after resharding and downloaded the state of shard B directly. Then it won't have the mapping from B to A.

That's quite interesting! It means two nodes may represent the same state differently depending on their shard tracking history. I don't think it's a problem, but it does feel like an unexpected consequence to me.

Reply:
Ah yes. Updated here: https://github.com/near/NEPs/pull/583